### PR TITLE
Adding Permutation Linear Operators

### PR DESCRIPTION
**Pull request by SebastianAment**
This PR adds `PermutationLinearOperator` and `TransposePermutationLinearOperator`. The first one represents a generic 2d permutation matrix, while the second one is a special type of permutation `P` that satisfies
```python
P @ X.flatten() = vec(X.transpose(-2, -1)).
```

This transpose permutation shows up in recent Monarch matrices, as well as the structured representations of kernel matrices of gradient kernels.

A caveat is that the test suites do not inherit from `LinearOperatorTestCase`, since that leads to a lot of errors mainly due to missing support for changing batch shapes. Especially for the transpose permutation, adding support for `expand` and `reshape` seems unnatural, since it would require the introduction of a `batch_shape` field, that is completely independent of the functioning of the operator except to make it play nice with the existing testing infrastructure. 

Let me know what you think!

### CODE REVIEW

File: linear_operator/operators/permutation_linear_operator.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    r"""LinearOperator that represents a permutation matrix `P` with O(1) memory.
+    In particular, P satisfies
+
+        `P @ X.flatten() = vec(X.transpose(-2, -1))`,
+
+    where `X` is an `m x m` matrix and P has size `n x n` where `n = m^2`.
+
+    Args:
+        - n: size of the permutation matrix that the operator represents.
+            NOTE: Has to be a perfect square, i.e. there is an `m` such that `n = m^2`.
```
Comment (SebastianAment) [inline linear_operator/operators/permutation_linear_operator.py]: ... there is an *integer* `m` such that ...

File: linear_operator/operators/permutation_linear_operator.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        return self.inverse() @ rhs
+
+    def _size(self):
+        n = len(self)
+        return (n, n)
+
+    def _check_shape(self, rhs) -> None:
+        if rhs.shape[-2] != len(self):
+            raise ValueError(
+                f"rhs.shape[-2] = {rhs.shape[-2]} incompatible " f"with size of permutation operator {len(self)}."
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]:
```suggestion
                f"rhs.shape[-2] = {rhs.shape[-2]} incompatible "
                f"with size of permutation operator {len(self)}."
```

File: linear_operator/operators/permutation_linear_operator.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        self,
+        perm: Tensor,
+        inv_perm: Optional[Tensor] = None,
+        check: bool = True,
+    ):
+        if not isinstance(perm, Tensor):
+            raise ValueError("perm is not a Tensor.")
+
+        if not perm.ndim == 1:
+            raise ValueError("perm is not a 1d Tensor.")
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]: Would we at some point want to support a batched permutation tensor? Say where `perm` is of size `(batch_shape) x n` where `perm[i, :]` indicates the permutation of the `i`-th batch (here `i` could be multidimensional as well).

File: linear_operator/operators/permutation_linear_operator.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        - inv_perm: Optional tensor representing the inverse of perm, is computed
+            via a O(n log(n)) sort if not given.
+        - check: Boolean
+    """
+
+    def __init__(
+        self,
+        perm: Tensor,
+        inv_perm: Optional[Tensor] = None,
+        check: bool = True,
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]: how about calling this `validate_args` to be consistent with the torch.distributions naming?

File: linear_operator/operators/permutation_linear_operator.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        if inv_perm is not None:
+            sorted_perm = perm[inv_perm]
+        else:
+            sorted_perm, inv_perm = perm.sort()
+
+        if check:
+            for i, si in enumerate(sorted_perm):
+                if torch.is_floating_point(si) or torch.is_complex(si):
+                    raise ValueError("perm does not have integer elements.")
+                if sorted_perm[i] != i:
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]: no need to do this if `inv_perm` is `None`.

File: linear_operator/operators/permutation_linear_operator.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            raise ValueError("perm is not a 1d Tensor.")
+
+        if inv_perm is not None:
+            sorted_perm = perm[inv_perm]
+        else:
+            sorted_perm, inv_perm = perm.sort()
+
+        if check:
+            for i, si in enumerate(sorted_perm):
+                if torch.is_floating_point(si) or torch.is_complex(si):
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]: You can call these directly on `perm` and `inv_perm` instead of on the individual elements

File: linear_operator/operators/permutation_linear_operator.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+                if sorted_perm[i] != i:
+                    raise ValueError(
+                        f"Invalid perm-inv_perm input, index {i} missing or not at "
+                        f"correct index for permutation of length {len(perm)}."
+                    )
+        self.perm = perm
+        self.inv_perm = inv_perm
+        super().__init__(perm, inv_perm, check=check)
+
+    def _matmul(self, rhs: Tensor):
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]:
```suggestion
    def _matmul(self, rhs: Tensor) -> Tensor:
```

File: linear_operator/operators/permutation_linear_operator.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        self._check_shape(rhs)
+        return rhs[..., self.perm, :]
+
+    def __len__(self):
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]:
```suggestion
    def __len__(self) -> int:
```

File: linear_operator/operators/permutation_linear_operator.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        return len(self.perm)
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]: Is defining this necessary here? Many `LinearOperator`s don't have a `__len__` defined, and the one here is inconsistent with the usual `__len__` behavior of dense tensors (where this would give the size of the leading dimension).

Comment (Balandat) [reply linear_operator/operators/permutation_linear_operator.py]: Seems like you should be able to just call `len(self.perm)` internally as needed in a few places

Comment (Balandat) [reply linear_operator/operators/permutation_linear_operator.py]: You could just use `.shape[-1]` here

File: linear_operator/operators/permutation_linear_operator.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        for (i, pi) in self.perm:
+            P[pi, i] = 1
+        return P
+
+
+class TransposePermutationLinearOperator(AbstractPermutationLinearOperator):
+    r"""LinearOperator that represents a permutation matrix `P` with O(1) memory.
+    In particular, P satisfies
+
+        `P @ X.flatten() = vec(X.transpose(-2, -1))`,
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]: If `X` were batched then `flatten()` would not work here so if you want batch support you could do this here
```suggestion
        `P @ X.flatten(-2, -1) = vec(X.transpose(-2, -1))`,
```

File: linear_operator/operators/permutation_linear_operator.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+    where `X` is an `m x m` matrix and P has size `n x n` where `n = m^2`.
+
+    Args:
+        - n: size of the permutation matrix that the operator represents. NOTE: `n` has
+            to be a perfect square, i.e. there is an integer `m` such that `n = m^2`.
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]: why not make `m` the arg to the constructor?

File: linear_operator/operators/permutation_linear_operator.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    def __init__(self, n: int):
+        m = isqrt(n)  # integer square root
+        if not m * m == n:
+            raise ValueError(f"n = {n} is not a perfect square")
+
+        self.n = n  # size of implicitly represented linear operator
+        self.m = m  # (m, m) is size of the reshaped input which is transposed
+        super().__init__(n=self.n)
+
+    def _matmul(self, rhs: Tensor):
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]:
```suggestion
    def _matmul(self, rhs: Tensor) -> Tensor:
```

File: linear_operator/operators/permutation_linear_operator.py (hunk 7 above)
Comment (Balandat) [reply linear_operator/operators/permutation_linear_operator.py]: Should this not also work when `rhs` is a `LinearOperator`?

File: linear_operator/operators/permutation_linear_operator.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        return PermutationLinearOperator(perm=self.inv_perm, inv_perm=self.perm, check=False)
+
+    def inverse(self):
+        return self._transpose_nonbatch()
+
+    def to_dense(self):
+        P = torch.zeros(self.shape, dtype=self.perm.dtype)
+        for (i, pi) in self.perm:
+            P[pi, i] = 1
+        return P
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]:
```suggestion
    def to_dense(self) -> Tensor:
        P = torch.zeros(self.shape, dtype=self.perm.dtype)
        P[..., self.perm, torch.arange(len(self.perm))] = 1
        return P
```

Or, if you wanted to be self-referential,
```
return self @ torch.diag_embed(torch.ones(self.shape[:-1]))
```

File: linear_operator/operators/permutation_linear_operator.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+    def _check_shape(self, rhs) -> None:
+        if rhs.shape[-2] != len(self):
+            raise ValueError(
+                f"rhs.shape[-2] = {rhs.shape[-2]} incompatible " f"with size of permutation operator {len(self)}."
+            )
+
+
+class PermutationLinearOperator(AbstractPermutationLinearOperator):
+    r"""LinearOperator that lazily represents a permutation matrix with O(n) memory.
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py:34]: SHould be more clear that this permutes just the first non-batch dim here

File: linear_operator/operators/permutation_linear_operator.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+    def _transpose_nonbatch(self):
+        return self
+
+    @property
+    def dtype(self):
+        return torch.int
+
+    @property
+    def device(self):
+        return torch.device
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]: which `device`? Is there a a "universal device" in torch for objects that don't actually have any (tensor) memory allocated? May also just return `None` here instead (which corresponds to the default device in terms of behavior)

File: test/operators/test_permutation_linear_operator.py [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+        n = 5
+        P = PermutationLinearOperator(torch.randperm(n))
+        for x in [
+            torch.randn(n),
+            torch.randn(n, 3),
+            torch.randn(2, n, 3),
+            torch.randn(7, 2, n, 3),
+        ]:
+            # computed inverse permutation field sorts the permutation
+            self.assertTrue(torch.equal(P.perm[P.inv_perm], torch.tensor(range(n))))
```
Comment (Balandat) [inline test/operators/test_permutation_linear_operator.py]:
```suggestion
            self.assertTrue(torch.equal(P.perm[P.inv_perm], torch.arange(n)))
```

File: linear_operator/operators/permutation_linear_operator.py (hunk 14 above)
Comment (SebastianAment) [reply linear_operator/operators/permutation_linear_operator.py:34]: I added "Upon left-multiplication, it permutes the first non-batch dimension of a tensor." in an upcoming commit.

File: linear_operator/operators/permutation_linear_operator.py (hunk 5 above)
Comment (SebastianAment) [reply linear_operator/operators/permutation_linear_operator.py]: If `inv_perm` is `None`, `sorted_perm` is guaranteed to be ordered correctly, but this can still fail if `perm` is not actually a permutation. I am testing this case with 
```python
with self.assertRaisesRegex(ValueError, "Invalid perm*"):
     PermutationLinearOperator(torch.tensor([1, 3, 5]))
```
in the test file. So I think there is still value in keeping this test. What do you think?

File: linear_operator/operators/permutation_linear_operator.py (hunk 7 above)
Comment (SebastianAment) [reply linear_operator/operators/permutation_linear_operator.py]: As far as I can tell, the `matmul` with a `LinearOperator` should always be represented lazily, which is the default behavior of `matmul` in `_linear_operator`. That means we'll only ever get `Tensor` types be passed into `_matmul`. As an example, Monarch matrices require the lazy `MatmulLinearOperator` representation of a `PermutationLinearOperator` and a `BlockDiagLinearOperator` to be computationally efficient. Is there another case I am forgetting about?

File: linear_operator/operators/permutation_linear_operator.py (hunk 13 above)
Comment (SebastianAment) [reply linear_operator/operators/permutation_linear_operator.py]: That's clean!

File: linear_operator/operators/permutation_linear_operator.py (hunk 11 above)
Comment (SebastianAment) [reply linear_operator/operators/permutation_linear_operator.py]: Done in the upcoming commit.

**Comment by SebastianAment** [comment]
@Balandat The latest commit supports batched `PermutationLinearOperators` that you proposed [here](https://github.com/cornellius-gp/linear_operator/pull/39#discussion_r1020427301). The indexing logic required to make this happen increased the complexity of the implementation.

For `TransposePermutationOperator`, I also introduced a `batch_shape` argument, but am not sure if we should keep it. It is a step towards unifying its interface with the rest of the `LinearOperators`, but is a primarily syntactic addition.

I replaced the custom `to_dense` method with a `to_sparse` method, since the `to_dense` is already covered by the base class implementation. However, the PyTorch's CSR tensor implementation only allows batch support for `torch.__version__ > 1.12`, so I am toggling the corresponding test based on the torch version. The test passes locally.

File: linear_operator/operators/permutation_linear_operator.py [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+    def _matmul_batch_shape(self, rhs: Tensor) -> torch.Size:
+        return torch.broadcast_shapes(self.batch_shape, rhs.shape[:-2])
+
+
+class PermutationLinearOperator(AbstractPermutationLinearOperator):
+    r"""LinearOperator that lazily represents a permutation matrix with O(n) memory.
+    Upon left-multiplication, it permutes the first non-batch dimension of a tensor.
+
+    Args:
+        - perm: A permutation tensor with which to permute the first non-batch
```
Comment (SebastianAment) [inline linear_operator/operators/permutation_linear_operator.py:38]: I should add information regarding batched `perm` tensors here.

File: linear_operator/operators/permutation_linear_operator.py [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+class TransposePermutationLinearOperator(AbstractPermutationLinearOperator):
+    r"""LinearOperator that represents a permutation matrix `P` with O(1) memory.
+    In particular, P satisfies
+
+        `P @ X.flatten(-2, -1) = X.transpose(-2, -1).flatten(-2, -1)`,
+
+    where `X` is an `m x m` matrix and P has size `n x n` where `n = m^2`.
+
+    Args:
+        - m: dimension on which the transpose operation is taking place. The size of
```
Comment (SebastianAment) [inline linear_operator/operators/permutation_linear_operator.py:136]: If we decide to keep `batch_shape`, I should add it to the doc string here.

File: linear_operator/operators/permutation_linear_operator.py [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+        return tuple(indices)
+
+    def _size(self) -> torch.Size:
+        return torch.Size((*self.perm.shape, self.perm.shape[-1]))
+
+    def _transpose_nonbatch(self):
+        return PermutationLinearOperator(perm=self.inv_perm, inv_perm=self.perm, validate_args=False)
+
+    def to_sparse(self):
+        # crow_indices[i] is index where values of row i begin
```
Comment (SebastianAment) [inline linear_operator/operators/permutation_linear_operator.py:119]: Missing doc-string at this point.

File: linear_operator/operators/permutation_linear_operator.py [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+
+        # batch broadcasting logic
+        batch_shape = self._matmul_batch_shape(rhs)
+        expanded_rhs = rhs.expand(*batch_shape, *rhs.shape[-2:])
+        ndim = expanded_rhs.ndim
+
+        batch_indices = self._batch_indexing_helper(batch_shape)
+        batch_indices = tuple(index[..., None] for index in batch_indices)  # expanding to non-batch dimensions
+        perm_indices = self.perm[..., None]
+        final_indices = torch.arange(rhs.shape[-1])[(None,) * (ndim - 1) + (slice(None),)]
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]: Not sure if cleaner but you could also do this
```suggestion
        final_indices = torch.arange(rhs.shape[-1]).view(*(1,)*(ndim-1), -1)
```

File: linear_operator/operators/permutation_linear_operator.py [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+        for i, n in enumerate(batch_shape):
+            indices.append(
+                torch.arange(n)[
+                    (None,) * i
+                    + (slice(None),)
+                    + (None,) * (len(batch_shape) - i - 1)
+                    + (None,)  # adding one non-batch dimension
+                ]
+            )
+        return tuple(indices)
```
Comment (Balandat) [inline linear_operator/operators/permutation_linear_operator.py]: Can make a tuple right away. Also could use `view` here instead (not necessary though)
```suggestion
        return tuple(
            torch.arange(n)[
                (None,) * i
                + (slice(None),)
                + (None,) * (len(batch_shape) - i - 1)
                + (None,)  # adding one non-batch dimension
            ]
            for i, n in enumerate(batch_shape)
        )
```

File: linear_operator/operators/permutation_linear_operator.py (hunk 18 above)
Comment (Balandat) [reply linear_operator/operators/permutation_linear_operator.py:136]: Hmm the only thing batch shape does is to give it a specific `size`? I don't think that's necessary so long as the operator is properly broadcastable (which it seems it is). Not sure if we want to ever represent the permutation with a specific batch shape. @gpleiss any opinions here?

**Comment by SebastianAment** [comment]
I changed the `None`-based indexing to `view`-based indexing, which I think improves readability. Thanks for the suggestion @Balandat ! 

I also removed `batch_shape` from `TransposePermutationLinearOperator`. Its only value is to unify its interface with other `LinearOperators` and allow e.g. `expand` calls. I'll put it back in if the need ever arises, but think we can ship this version first.

If you approve the current version, could you hold off on merging directly? I would like to squash the commits before.

**Comment by Balandat** [comment]
> If you approve the current version, could you hold off on merging directly? I would like to squash the commits before.

Sure, please go ahead. Good to keep things tidy :)

**Review by Balandat — APPROVED** [review]

**Comment by SebastianAment** [comment]
Good to go! :)

**Comment by SebastianAment** [comment]
One unexpectedly adventurous rebase later, this really is now good to go.
